execution/tvf/percentile_watermark_generator.go

```diff
@@ -0,0 +1,297 @@
+	timeField  octosql.VariableName
+	events     execution.Expression
+	percentile execution.Expression
+	frequency  execution.Expression
+}
+
+func NewPercentileWatermarkGenerator(source execution.Node, timeField octosql.VariableName, events, percentile, frequency execution.Expression) *PercentileWatermarkGenerator {
+	return &PercentileWatermarkGenerator{
+		source:     source,
+		timeField:  timeField,
+		events:     events,
+		percentile: percentile,
+		frequency:  frequency,
+	}
+}
+
+func (w *PercentileWatermarkGenerator) Get(ctx context.Context, variables octosql.Variables, streamID *execution.StreamID) (execution.RecordStream, *execution.ExecOutput, error) {
+	tx := storage.GetStateTransactionFromContext(ctx)
+	sourceStreamID, err := execution.GetSourceStreamID(tx.WithPrefix(streamID.AsPrefix()), octosql.MakePhantom())
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "couldn't get source stream ID")
+	}
+
+	source, _, err := w.source.Get(ctx, variables, sourceStreamID) // we don't need execOutput here since we become new watermark source
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "couldn't get source")
+	}
+
+	events, err := w.events.ExpressionValue(ctx, variables)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "couldn't get watermark events")
+	}
+	if events.GetType() != octosql.TypeInt || events.AsInt() < 1 {
+		return nil, nil, errors.Errorf("invalid watermark events: %v", events)
+	}
+
+	percentile, err := w.percentile.ExpressionValue(ctx, variables)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "couldn't get watermark percentile")
+	}
+	if percentile.GetType() != octosql.TypeInt || percentile.AsInt() < 1 || percentile.AsInt() > 99 {
+		return nil, nil, errors.Errorf("invalid watermark percentile: %v", percentile)
+	}
+
+	frequency, err := w.frequency.ExpressionValue(ctx, variables)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "couldn't get watermark frequency")
+	}
+	if frequency.GetType() != octosql.TypeInt || frequency.AsInt() < 1 {
+		return nil, nil, errors.Errorf("invalid watermark frequency: %v", frequency)
+	}
+
+	ws := &PercentileWatermarkGeneratorStream{
+		source:     source,
+		timeField:  w.timeField,
+		events:     events.AsInt(),
+		percentile: percentile.AsInt(),
+		frequency:  frequency.AsInt(),
+	}
+
+	return ws, execution.NewExecOutput(ws), nil // watermark generator stream now indicates new watermark source
+}
+
+// The way this watermark generator is working is following:
+// - events (must be positive) - represents amount of (most recent) events stored
+// - percentile (must be between 1 and 99) - represents percentile of recently stored events that are BIGGER than watermark value
+// 		ex. if percentile = 35 then 35% of events stored must be bigger than watermark value, so watermark position is at
+//			65th percentile of events stored (in sorted way)
+// - frequency (must be positive) - represents amount of events to be seen to initiate updating watermark process
```

Comment:
This comment line sounds like it starts after "frequency" records and later goes every one record. Maybe "represents amount of events to be seen before initiating next watermark update"?